src/withdrawal_limit/component.cairo

```diff
@@ -0,0 +1,126 @@
+#[starknet::component]
+pub mod WithdrawalLimitComponent {
+    use starknet::{ContractAddress, get_block_timestamp};
+    use starknet_bridge::bridge::interface::ITokenBridge;
+    use starknet_bridge::constants;
+    use core::integer::BoundedInt;
+
+    use openzeppelin::token::erc20::interface::{IERC20Dispatcher, IERC20DispatcherTrait};
+    use starknet_bridge::withdrawal_limit::interface::IWithdrawalLimit;
+
+    #[storage]
+    struct Storage {
+        // For each token and day, stores the amount that can still be withdrawn from this token
+        // in this day (if the value is x, the amount left to withdraw is x-1). 0 means that
+        // currently there was no withdrawal from this token in this day or there were withdrawals
+        // but the limit flag was turned off.
+        remaining_intraday_withdraw_quota: LegacyMap<(ContractAddress, u64), u256>,
+        // The daily withdrawal limit percentage.
+        daily_withdrawal_limit_pct: u8,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        RemainingQuotaUpdated: RemainingQuotaUpdated
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct RemainingQuotaUpdated {
+        new_quota: u256
+    }
+
+
+    #[embeddable_as(WithdrawalLimitImpl)]
+    pub impl WithdrawalLimit<
+        TContractState, +HasComponent<TContractState>, +ITokenBridge<TContractState>
+    > of IWithdrawalLimit<ComponentState<TContractState>> {
+        // Returns the current remaining withdrawal quota for a given token. If there is no limit,
+        // returns max uint256. If the limit was not set yet, we calculate it based on the total
+        // supply. Otherwise, return the limit.
+        fn get_remaining_withdrawal_quota(
+            self: @ComponentState<TContractState>, token: ContractAddress
+        ) -> u256 {
+            // If there is no limt, return max uint256.
+            if self.get_contract().is_withdrawal_limit_applied(:token) == false {
```

Comment:
i like the L1 design better. the `WithdrawalLimit` library assumes that its functions are only called if the limit is enabled (checking that is the callers responsibility). but once the code is inside the library, we can assume that it is indeed enabled. our current code looks a little mixed, some places we check for limit, somewhere we check for uint max. am i missing something?

Reply:
- Components in cairo embed into the contract and do create external functions unlike the library in solidity.
- I think `get_remaining_withdrawal_quota()` makes sense to be an external function here to be called by users
- To check the if `withdrawal_limit` is applied or not there were 2 ways either replicate the storage in the component or call the function in the contract using `get_contract()`

I also get that this is making the component non pluggable in other contracts, I now like the idea of changing the trait bound from `+ITokenBridge<TContractState>` and define a new interface say IWithdrawalLimitStatus and use `+IWithdrawalLimitStatus<TContractStatus>` as the trait bound. The interface would define just one function `is_withdrawal_limit_applied()`